src/components/transactions/Repay/CollateralRepayModalContent.tsx

```diff
@@ -141,22 +193,50 @@ export function CollateralRepayModalContent({
     }
   };
 
-  // for v3 we need hf after withdraw collateral, because when removing collateral to repay
-  // debt, hf could go under 1 then it would fail. If that is the case then we need
-  // to use flashloan path
+  // health factor calculations for aToken repayments
+  // we use usd values instead of MarketreferenceCurrency so it has same precision
+  let newHFWithATokens = user?.healthFactor;
+  if (amount && repayWithATokens) {
+    let collateralBalanceMarketReferenceCurrency: BigNumberValue = user?.totalCollateralUSD || '0';
+    if (repayWithATokens && usageAsCollateralEnabledOnUser) {
+      collateralBalanceMarketReferenceCurrency = valueToBigNumber(
+        user?.totalCollateralUSD || '0'
+      ).minus(valueToBigNumber(reserve.priceInUSD).multipliedBy(amount));
+    }
+
+    const remainingBorrowBalance = valueToBigNumber(user?.totalBorrowsUSD || '0').minus(
+      valueToBigNumber(reserve.priceInUSD).multipliedBy(amount)
+    );
+    const borrowBalanceMarketReferenceCurrency = BigNumber.max(remainingBorrowBalance, 0);
+
+    const calculatedHealthFactor = calculateHealthFactorFromBalancesBigUnits({
+      collateralBalanceMarketReferenceCurrency,
+      borrowBalanceMarketReferenceCurrency,
+      currentLiquidationThreshold: user?.currentLiquidationThreshold || '0',
+    });
+
+    newHFWithATokens =
+      calculatedHealthFactor.isLessThan(0) && !calculatedHealthFactor.eq(-1)
+        ? '0'
+        : calculatedHealthFactor.toString(10);
+  }
+
```

Comment:
Wonder if it'd be possible to use the `calculateHFAFterRepay` function for the aToken case?